switchio/loops/proactor.py

```diff
@@ -0,0 +1,330 @@
+            # block waiting for next event
+            e = await self._rx_con.recv_event()
+            # self.log.warning(get_event_time(e) - self._fs_time)
+            if e is None:
+                self.log.debug("Exiting listen loop")
+                break
+            elif not e:
+                self.log.error("Received empty event!?")
+            else:
+                evname = e.get('Event-Name')
+                if evname:
+                    consumed = await self._process_event(e, evname)
+                    if not consumed:
+                        self.log.warn("unconsumed  event '{}'?".format(e))
+                else:
+                    self.log.warn("received unnamed event '{}'?".format(e))
+
+        self.log.debug("Exiting listen loop")
+        self._running = False
+
+    async def _process_event(self, e, evname):
+        '''Process an ESL event by delegating to the appropriate handler
+        and any succeeding callback chain. This is the core handler lookup
+        routine and should be optimized for speed.
+
+        An event is considered consumed if:
+        1) the handler + callback chain returns True
+        2) the handler + callback chain raises a special exception
+
+        :param ESL.ESLEvent e: event received over esl on self._rx_con
+        :param str evname: event type/name string
+        '''
+        # epoch is the time when first event is received
+        if self._epoch:
+            self._fs_time = get_event_time(e)
+        else:
+            self._epoch = self._fs_time = get_event_time(e)
+
+        consumed = False  # is this event consumed by a handler/callback
+        if 'CUSTOM' in evname:
+            evname = e.get('Event-Subclass')
+        self.log.debug("receive event '{}'".format(evname))
+
+        uid = e.get('Unique-ID')
+
+        handler = self._handlers.get(evname, False)
+        loop = self.loop
+        if handler:
+            self.log.debug("handler is '{}'".format(handler))
+            try:
+                consumed, ret = utils.uncons(*handler(e))  # invoke handler
+                model = ret[0]
+
+                # attempt to lookup a consuming client app (callbacks) by id
+                cid = model.cid if model else self.get_id(e, 'default')
+                self.log.debug("app id is '{}'".format(cid))
+
+                callbacks = self.callbacks.get(cid, False)
+                if callbacks and consumed:
+                    cbs = callbacks.get(evname, ())
+                    self.log.debug(
+                        "consumer '{}' has callback {} registered for ev {}"
+                        .format(cid, cbs, evname)
+                    )
+                    # look up the client's callback chain and run
+                    # e -> handler -> cb1, cb2, ... cbN
+                    # XXX assign ret on each interation in an attempt to avoid
+                    # python's dynamic scope lookup
+                    for cb, ret in zip(cbs, itertools.repeat(ret)):
```

Comment:
Might want to look at using `itertools.izip` for python 2, depending how much you care about that platform at this point.